### PR TITLE
feat: Add workspace proxy enterprise cli commands

### PR DESCRIPTION
**Pull request by Emyrk**
# What this does

Makes a cli command to spin up external workspace proxies!

```
coder proxy register # ....
coder proxy server # ....
```

## Testing/Development

This gives `develop.sh` an option `--use-proxy` which will spin up the devserver with an external workspace proxy. This should allow easier manual validation.

## Features added to make this work

- Cli command to delete workspace proxies.
- `clibase.Optionset` filter to filter out options that do not pertain to the workspace proxy.
- Ability to hard code `app-security-key`

# Logs

Run `./develop.sh --use-proxy`

```
====================================================================
==                                                                ==
==            Coder is now running in development mode.           ==
==                  API:    http://localhost:3000                 ==
==                  API:    http://127.0.0.1:3000                 ==
==                  API:    http://192.168.1.17:3000              ==
==                  API:    http://192.168.1.18:3000              ==
==                  API:    http://172.25.0.1:3000                ==
==                  Web UI: http://localhost:8080                 ==
==                  Web UI: http://127.0.0.1:8080                 ==
==                  Web UI: http://192.168.1.17:8080              ==
==                  Web UI: http://192.168.1.18:8080              ==
==                  Web UI: http://172.25.0.1:8080                ==
==                  Proxy:  http://localhost:3010                 ==
==                  Proxy:  http://127.0.0.1:3010                 ==
==                  Proxy:  http://192.168.1.17:3010              ==
==                  Proxy:  http://192.168.1.18:3010              ==
==                  Proxy:  http://172.25.0.1:3010                ==
==                                                                ==
==      Use ./scripts/coder-dev.sh to talk to this instance!      ==
==       alias cdr=/home/steven/go/src/github.com/cdr/coder/scripts/coder-dev.sh              ==
====================================================================
```

### CODE REVIEW

File: cli/appsecuritykey.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/coder/coder/cli/clibase"
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/coderd/database/migrations"
+)
+
+func (*RootCmd) appSecurityKey() *clibase.Cmd {
+	var postgresURL string
+
+	root := &clibase.Cmd{
+		Use:   "app-security-key",
```
Comment (deansheather) [inline cli/appsecuritykey.go]: We should change this so the moon fetches the key on startup. That can be part of the registration issue though.

File: coderd/httpmw/workspaceproxy.go [hunk 2]
```diff
@@ -8,6 +8,8 @@ import (
 	"net/http"
 	"strings"
 
+	"github.com/go-chi/chi/v5"
+
 	"github.com/google/uuid"
 	"golang.org/x/xerrors"
```
Comment (deansheather) [inline coderd/httpmw/workspaceproxy.go:15]: import ordering

File: enterprise/cli/workspaceproxy.go [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+
+			_, _ = fmt.Fprintf(inv.Stdout, "Workspace proxy %q deleted successfully\n", inv.Args[0])
+			return nil
+		},
+	}
+
+	return cmd
+}
+
+func (r *RootCmd) registerProxy() *clibase.Cmd {
```
Comment (deansheather) [inline enterprise/cli/workspaceproxy.go:78]: We should call it create TBH, the "registration" process will happen using a request from the moon with the token.

File: enterprise/cli/workspaceproxy.go [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+	cmd := &clibase.Cmd{
+		Use:     "server",
+		Short:   "Start a workspace proxy server",
+		Options: opts,
+		Middleware: clibase.Chain(
+			cli.WriteConfigMW(cfg),
+			cli.PrintDeprecatedOptions(),
+			clibase.RequireNArgs(0),
+			// We need a client to connect with the primary coderd instance.
+			r.InitClient(client),
```
Comment (deansheather) [inline enterprise/cli/workspaceproxy.go:257]: Shouldn't this just be a proxysdk client that we construct in the handler?

File: enterprise/cli/workspaceproxy_slim.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/coder/coder/cli/clibase"
+	"github.com/coder/coder/cli/cliui"
+)
+
+func (r *RootCmd) workspaceProxy() *clibase.Cmd {
```
Comment (deansheather) [inline enterprise/cli/workspaceproxy_slim.go:14]: Only the server subcommand should be shimmed :/

File: enterprise/coderd/workspaceproxy.go [hunk 6]
```diff
@@ -7,6 +7,8 @@ import (
 	"net/http"
 	"net/url"
 
+	"github.com/coder/coder/coderd/httpmw"
```
Comment (deansheather) [inline enterprise/coderd/workspaceproxy.go:10]: import order

File: enterprise/cli/workspaceproxy.go [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+
+		externalProxyOptionGroup = clibase.Group{
+			Name: "External Workspace Proxy",
+			YAML: "externalWorkspaceProxy",
+		}
+		proxySessionToken clibase.String
+		primaryAccessURL  clibase.URL
+		appSecuritYKey    clibase.String
+	)
+	opts.Add(
```
Comment (deansheather) [inline enterprise/cli/workspaceproxy.go:205]: Can't these be added to the regular config struct, and then the primary server can filter them out?

File: cli/appsecuritykey.go (hunk 1 above)
Comment (Emyrk) [reply cli/appsecuritykey.go]: Agreed, removed this cli cmd.

File: enterprise/cli/workspaceproxy.go (hunk 4 above)
Comment (Emyrk) [reply enterprise/cli/workspaceproxy.go]: I was honestly just using this for the version checking.

File: enterprise/cli/workspaceproxy.go (hunk 7 above)
Comment (Emyrk) [reply enterprise/cli/workspaceproxy.go]: We talked on discord. I think the best move here is to autogenerate the Deployment struct from the Options. Then we can create all the options in 1 place, and filter them.

I will do this in another branch, this is _ok for now_ here just to get a cmd working.

Follow up PRs will make the workspace proxy creation/register experience better.

**Comment by Emyrk** [comment]
Rebased to https://github.com/coder/coder/pull/7123

**Pull request closed without merging**